crates/bevy_ui/src/stack.rs

```diff
@@ -15,102 +15,78 @@ pub struct UiStack {
-/// Then flatten that tree into back-to-front ordered `UiStack`.
 pub fn ui_stack_system(
     mut ui_stack: ResMut<UiStack>,
-    root_node_query: Query<Entity, (With<Node>, Without<Parent>)>,
-    zindex_query: Query<&ZIndex, With<Node>>,
+    root_node_query: Query<
+        (Entity, Option<&GlobalZIndex>, Option<&ZIndex>),
+        (With<Node>, Without<Parent>),
+    >,
+    zindex_global_node_query: Query<
+        (Entity, &GlobalZIndex, Option<&ZIndex>),
+        (With<Node>, With<Parent>),
+    >,
     children_query: Query<&Children>,
+    zindex_query: Query<Option<&ZIndex>, (With<Node>, Without<GlobalZIndex>)>,
 ) {
-    // Generate `StackingContext` tree
-    let mut global_context = StackingContext::default();
-    let mut total_entry_count: usize = 0;
-
-    for entity in &root_node_query {
-        insert_context_hierarchy(
-            &zindex_query,
-            &children_query,
-            entity,
-            &mut global_context,
-            None,
-            &mut total_entry_count,
-        );
-    }
-
-    // Flatten `StackingContext` into `UiStack`
     ui_stack.uinodes.clear();
-    ui_stack.uinodes.reserve(total_entry_count);
-    fill_stack_recursively(&mut ui_stack.uinodes, &mut global_context);
-}
-
-/// Generate z-index based UI node tree
-fn insert_context_hierarchy(
-    zindex_query: &Query<&ZIndex, With<Node>>,
-    children_query: &Query<&Children>,
-    entity: Entity,
-    global_context: &mut StackingContext,
-    parent_context: Option<&mut StackingContext>,
-    total_entry_count: &mut usize,
-) {
-    let mut new_context = StackingContext::default();
-
-    if let Ok(children) = children_query.get(entity) {
-        // Reserve space for all children. In practice, some may not get pushed since
-        // nodes with `ZIndex::Global` are pushed to the global (root) context.
-        new_context.entries.reserve_exact(children.len());
-
-        for entity in children {
-            insert_context_hierarchy(
-                zindex_query,
-                children_query,
-                *entity,
-                global_context,
-                Some(&mut new_context),
-                total_entry_count,
-            );
+    let uinodes = &mut ui_stack.uinodes;
+
+    fn update_uistack_recursively(
+        entity: Entity,
+        uinodes: &mut Vec<Entity>,
+        children_query: &Query<&Children>,
+        zindex_query: &Query<Option<&ZIndex>, (With<Node>, Without<GlobalZIndex>)>,
+    ) {
+        uinodes.push(entity);
+
+        if let Ok(children) = children_query.get(entity) {
+            let mut z_children: Vec<(Entity, i32)> = children
+                .iter()
+                .filter_map(|entity| {
+                    zindex_query
+                        .get(*entity)
+                        .ok()
+                        .map(|zindex| (*entity, zindex.map(|zindex| zindex.0).unwrap_or(0)))
+                })
+                .collect();
```

Comment:
This potentially allocates and deallocates for every descendant in the UI tree. Can we use a `&mut Vec<(Entity, i32)>` as a scratch space instead of using collect here?

Reply:
I tried this but introducing a scratch space seemed to make the system slower, even after switching to using iterative stack based tree traversal. I think this PR is good enough as it is without it anyway.